holoviews/plotting/bokeh/element.py

```diff
@@ -190,9 +190,18 @@ def _construct_callbacks(self):
         Initializes any callbacks for streams which have defined
         the plotted object as a source.
         """
+        if isinstance(self, OverlayPlot):
+            zorders = []
+        elif self.batched:
+            zorders = list(range(self.zorder, self.zorder+len(self.hmap.last)))
+        else:
+            zorders = [self.zorder]
+
+        if isinstance(self, OverlayPlot) and not self.batched:
+            sources = []
         if not self.static or isinstance(self.hmap, DynamicMap):
-            sources = [(i, o) for i, o in get_sources(self.hmap)
-                       if i in [None, self.zorder]]
+            sources = [(i, o) for i, inputs in self.stream_sources.items()
+                       for o in inputs if i in zorders]
```

Comment:
What does being a source have to do with a z-ordering? The latter is a visualization concept so what does that have to with being a source or not?